File: frontend/libs/data-access/src/lib/subscription.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse, PaginatedData, Subscription } from './models';
import { environment } from './environment';

export interface SubscriptionPayload {
  company_id: number;
  name: string;
  plan_type: string;
  status?: string;
  amount: number;
  currency?: string;
  billing_cycle?: string;
  start_date: string;       // YYYY-MM-DD
  renewal_date?: string | null;
  notes?: string;
}

@Injectable({ providedIn: 'root' })
export class SubscriptionService {
  private http = inject(HttpClient);
  private base = `${environment.apiUrl}/subscriptions`;

  getAll(offset = 0, limit = 50): Observable<ApiResponse<PaginatedData<Subscription>>> {
    return this.http.get<ApiResponse<PaginatedData<Subscription>>>(
      `${this.base}?offset=${offset}&limit=${limit}`
    );
  }

  getExpiringSoon(days = 30): Observable<ApiResponse<{ list: Subscription[] }>> {
    return this.http.get<ApiResponse<{ list: Subscription[] }>>(
      `${this.base}/expiring?days=${days}`
    );
  }

  getById(id: number): Observable<ApiResponse<{ subscription: Subscription }>> {
    return this.http.get<ApiResponse<{ subscription: Subscription }>>(`${this.base}/${id}`);
  }

  create(payload: SubscriptionPayload): Observable<ApiResponse<{ subscription: Subscription }>> {
    return this.http.post<ApiResponse<{ subscription: Subscription }>>(this.base, payload);
  }

  update(id: number, payload: SubscriptionPayload): Observable<ApiResponse<{ subscription: Subscription }>> {
    return this.http.put<ApiResponse<{ subscription: Subscription }>>(`${this.base}/${id}`, payload);
  }

  delete(id: number): Observable<ApiResponse<{ message: string }>> {
    return this.http.delete<ApiResponse<{ message: string }>>(`${this.base}/${id}`);
  }
}

File: backend/internal/core/services/subscription_service.go
package services

import (
	"errors"

	"github.com/sparkbigs/crm/internal/core/domain"
	"github.com/sparkbigs/crm/internal/core/ports"
)

var (
	ErrSubscriptionNotFound  = errors.New("suscripción no encontrada")
	ErrSubscriptionForbidden = errors.New("no tienes permiso sobre esta suscripción")
)

type subscriptionService struct {
	repo ports.SubscriptionRepository
}

func NewSubscriptionService(repo ports.SubscriptionRepository) ports.SubscriptionService {
	return &subscriptionService{repo: repo}
}

func (s *subscriptionService) CreateSubscription(sub *domain.Subscription) error {
	return s.repo.Create(sub)
}

func (s *subscriptionService) GetSubscription(id, userID uint) (*domain.Subscription, error) {
	sub, err := s.repo.FindByID(id)
	if err != nil {
		return nil, ErrSubscriptionNotFound
	}
	if sub.UserID != userID {
		return nil, ErrSubscriptionForbidden
	}
	return sub, nil
}

func (s *subscriptionService) ListSubscriptions(userID uint, offset, limit int) ([]domain.Subscription, int64, error) {
	return s.repo.FindByUserID(userID, offset, limit)
}

func (s *subscriptionService) ExpiringSoon(userID uint, days int) ([]domain.Subscription, error) {
	return s.repo.FindExpiringSoon(userID, days)
}

func (s *subscriptionService) UpdateSubscription(sub *domain.Subscription, userID uint) error {
	existing, err := s.repo.FindByID(sub.ID)
	if err != nil {
		return ErrSubscriptionNotFound
	}
	if existing.UserID != userID {
		return ErrSubscriptionForbidden
	}
	existing.Name         = sub.Name
	existing.PlanType     = sub.PlanType
	existing.Status       = sub.Status
	existing.Amount       = sub.Amount
	existing.Currency     = sub.Currency
	existing.BillingCycle = sub.BillingCycle
	existing.StartDate    = sub.StartDate
	existing.RenewalDate  = sub.RenewalDate
	existing.Notes        = sub.Notes
	existing.CompanyID    = sub.CompanyID
	return s.repo.Update(existing)
}

func (s *subscriptionService) DeleteSubscription(id, userID uint) error {
	existing, err := s.repo.FindByID(id)
	if err != nil {
		return ErrSubscriptionNotFound
	}
	if existing.UserID != userID {
		return ErrSubscriptionForbidden
	}
	return s.repo.Delete(id)
}
